### PR TITLE
Retry nightly test failures

### PR DESCRIPTION
**Pull request by jonsimantov**
On nightly tests, if any builds or tests failed, trigger a workflow that runs a script to retry failed jobs in the integration test run if appropriate.

The jobs will NOT be retried if:
- They are already in progress or queued (e.g. a retry is already happening)
- They have already been attempted 3 times 
- They contain a real compiler error

Otherwise, failed jobs WILL be retried if:
- They failed on mobile
- On desktop, the logs match retryable error text (e.g. "timed out", "network error")

This PR also includes some small fixes in trigger_workflow.py to fix issues brought up by the workflow changes:
- Don't try to start a workflow run at 'HEAD', use 'main' instead.
- Fix exit codes to actually work
- If git can't get the branch sha (for example if the branch name is specified manually and isn't known locally), ignore branch_sha when finding the started workflow.
- If a single workflow run can't be identified, print a human-friendly URL to click on to list running workflows rather than the API URL

***

Tested in this PR with the workflow set up to allow it to run outside of nightly tests (for debugging); this has been removed and now it will only run on nightlies.
***

### Type of Change
Place an `x` the applicable box:
- [ ] Bug fix. Add the issue # below if applicable.
- [ ] New feature. A non-breaking change which adds functionality.
- [x] Other, such as a build process or documentation change.
***

#### Notes
- Bug fixes and feature changes require an update to the `Release Notes` section of `release_build_files/readme.md`.
- Read the contribution guidelines [CONTRIBUTING.md](https://github.com/firebase/firebase-cpp-sdk/blob/main/CONTRIBUTING.md).
- Changes to the public API require an internal API review. If you'd like to help us make Firebase APIs better, please propose your change in a feature request so that we can discuss it together.

### CODE REVIEW

File: .github/workflows/integration_tests.yml [hunk 1]
```diff
@@ -1412,3 +1412,33 @@ jobs:
...
+      - name: Generate token for GitHub API
+        uses: tibdex/github-app-token@v1
+        id: generate-token
+        with:
+          app_id: ${{ secrets.WORKFLOW_TRIGGER_APP_ID }}
+          private_key: ${{ secrets.WORKFLOW_TRIGGER_APP_PRIVATE_KEY }}
+      - name: Retry failed tests
+        run: |
+          echo "::warning ::Attempting to retry failed tests"
+          python scripts/gha/trigger_workflow.py -t ${{ steps.generate-token.outputs.token }} \
```
Comment (a-maurice) [inline .github/workflows/integration_tests.yml:1441]: What workflow is this meant to be triggering?  It seems to be missing the -w flag

File: scripts/gha/retry_test_failures.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+  failed_jobs = {}
+  all_jobs = {}
+  for job in workflow_jobs:
+    all_jobs[job['id']] = job
+    if job['conclusion'] != 'success' and job['conclusion'] != 'skipped':
+      if job['name'] in failed_jobs:
+        other_run = failed_jobs[job['name']]
+        if job['run_attempt'] > other_run['run_attempt']:
+          # This is a later run than the one that's arleady there
```
Comment (a-maurice) [inline scripts/gha/retry_test_failures.py:81]: already

Comment (jonsimantov) [reply scripts/gha/retry_test_failures.py]: Fixed.

File: .github/workflows/integration_tests.yml (hunk 1 above)
Comment (jonsimantov) [reply .github/workflows/integration_tests.yml:1441]: Oops, fixed. Accidentally removed it when I removed the hard-coded branch for debugging.

**Review by a-maurice — APPROVED** [review]

**Pull request merged**